### PR TITLE
Fix: Lambda layer fails to upload to storage bucket (NoSuchBucket) (#3639)

### PR DESCRIPTION
**Pull request by achave11**
https://github.com/DataBiosphere/azul/issues/3639

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks   <sub>or this PR does not require additional tasks</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Cleanup old buckets used for deployment layers
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate       <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab

### Operator (reindex) 

- [ ] Started reindex in target deployment                  <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures             <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment              <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by achave11** [comment]
- [ ] I don't this will be a regression from the [previous lambda_layer_bucket](https://github.com/DataBiosphere/azul/pull/3450#discussion_r715035530) change, which addressed issue (https://github.com/DataBiosphere/azul/issues/3330). Just to confirm.
> [Also note that a lambda layer should not be deployed when make -C terraform plan is ran.](https://github.com/DataBiosphere/azul/issues/3639#issuecomment-965740447)
- [ ] Because `make -C terraform config` called by `… terraform plan` has baked in the notion that the lambdas have to be staged when running Terraform, that includes `make package` which deploys the lambdas. Should there be an alternate logic path which doesn't upload the layers to s3, but still produces the lambda resources with chalice-terraform?

**Comment by achave11** [comment]
add FIXME for `make -C terraform plan` uploads layers to s3.

```diff
Index: lambdas/Makefile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/Makefile b/lambdas/Makefile
--- a/lambdas/Makefile	(revision 39e1643b2804291e611017e5b82b8440aa0b2ead)
+++ b/lambdas/Makefile	(date 1637268693671)
@@ -1,5 +1,5 @@
 .PHONY: all
-all: layer indexer service
+all: indexer service
 
 include ../common.mk
 
Index: terraform/Makefile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/terraform/Makefile b/terraform/Makefile
--- a/terraform/Makefile	(revision 39e1643b2804291e611017e5b82b8440aa0b2ead)
+++ b/terraform/Makefile	(date 1637268693679)
@@ -27,11 +27,11 @@
 	terraform plan
 
 .PHONY: apply
-apply: validate rename_resources
+apply: validate layer rename_resources
 	terraform apply
 
 .PHONY: auto_apply
-auto_apply: validate rename_resources
+auto_apply: validate layer rename_resources
 	terraform apply -auto-approve
 
 .PHONY: destroy
@@ -67,3 +67,7 @@
 .PHONY: lambdas
 lambdas:
 	$(MAKE) -C $(project_root)/lambdas
+
+.PHONY: layer
+layer:
+	$(MAKE) -C $(project_root)/lambdas/layer
\ No newline at end of file

```

**Review by amarjandu — APPROVED** [review]
✅ Storage bucket is changed
✅ FIXME: added for weird relation between `plan` and layer upload

**Comment by amarjandu** [comment]
@achave11 might want to make a checklist item for the operator to cleanup the old bucket that was used.

File: terraform/Makefile [hunk 1]
```diff
@@ -8,6 +8,8 @@ providers: check_terraform check_branch check_aws providers.tf.json backend.tf.j
 	terraform init
 
 .PHONY: config
+# FIXME: lambda layers shouldn't be uploaded when `plan` target is ran
```
Comment (hannes-ucsc) [inline terraform/Makefile:11]:
```suggestion
# FIXME: Lambda layers shouldn't be uploaded when `plan` target is run
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Pull request merged**